master/internal/rm/agentrm/resource_pool.go

```diff
@@ -473,19 +475,42 @@ func (rp *resourcePool) Receive(ctx *actor.Context) error {
 	case sproto.CapacityCheck:
 		reschedule = false
 		var totalSlots int
+
+		disallowedNodes := set.New[string]()
+		if msg.TaskID != nil {
+			disallowedNodes = *logpattern.DisallowedNodes(*msg.TaskID)
+		}
+		rp.agentStatesCache = rp.fetchAgentStates(ctx)
+
 		switch {
 		case rp.config.Provider == nil:
-			rp.agentStatesCache = rp.fetchAgentStates(ctx)
+
 			defer func() {
```

Comment:
move this defer to outside of the switch so we are consistent about unsetting the cache